### PR TITLE
Store trielog as blobdb

### PR DESCRIPTION
**Pull request by usmansaleem**
## PR description
Store trielog as blobdb to improve performance. #5475 has shown that blobdb is better choice for static data as it reduced the rocksdb write amplifications during compaction. Manually tested upgrade and downgrade.

Graph showing bytes written during compaction. The yellow bar represents the node with "trielog as blobdb" PR applied (blobdb-2), the green bar represents the node without the blobdb patch (blobdb-3). On average, bytes written during compaction is lower for "trielog as blobdb vs trielog w/o blobdb":

image
--
Following graph shows fork choice update time (24 hours). There doesn't seem to be any negative impact of writing trielog as blobdb.

image

## Fixed Issue(s)

See https://github.com/hyperledger/besu/issues/5866

### CODE REVIEW

**Comment by siladu** [comment]
> If my PR includes database changes (e.g. KeyValueSegmentIdentifier) I have thought about compatibility and performed forwards and backwards compatibility tests

How was this part tested?

**Comment by siladu** [comment]
@usmansaleem I'm not really sure what the graph in the description is showing, can you add some explanation please?

Also, have a feeling we discussed it last year, but can't remember what the outcome of the following test was?
> Verify if pruning has an impact on performance due to compaction https://github.com/hyperledger/besu/issues/5866

**Comment by siladu** [comment]
> @usmansaleem I'm not really sure what the graph in the description is showing, can you add some explanation please?

ah, just spotted https://github.com/hyperledger/besu/issues/5866#issuecomment-1899360419

Have you got a graph for the trie log pruning enabled?

---
Last one...

>  Should we have a feature flag for this or should it be enabled by default?

I guess the answer to this was yes?

**Comment by jframe** [comment]
> Store trielog as blobdb to improve performance. https://github.com/hyperledger/besu/pull/5475 has shown that blobdb is better choice for static data as it reduced the rocksdb write amplifications during compaction. Manually tested upgrade and downgrade.
Do you have a graph on the impact on the engine API FCU and new payload?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/storage/keyvalue/KeyValueSegmentIdentifier.java [hunk 1]
```diff
@@ -28,7 +28,7 @@ public enum KeyValueSegmentIdentifier implements SegmentIdentifier {
   CODE_STORAGE(new byte[] {7}, new int[] {2}),
   ACCOUNT_STORAGE_STORAGE(new byte[] {8}, new int[] {2}),
   TRIE_BRANCH_STORAGE(new byte[] {9}, new int[] {2}),
-  TRIE_LOG_STORAGE(new byte[] {10}, new int[] {2}),
+  TRIE_LOG_STORAGE(new byte[] {10}, new int[] {2}, true),
```
[the remark this answers is not in the document]
Comment (fab-10) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/storage/keyvalue/KeyValueSegmentIdentifier.java:31]: No, RocksDB will start using blobs on new files only

**Review by fab-10 — CHANGES_REQUESTED** [review]
Since after https://github.com/hyperledger/besu/pull/6445 there can be deletions to this column family, there is the need to enable, and tune, blob garbage collection for it, since now it is it disabled by default (blockchain data cf is meant to be append only with no deletions*)

* there is an experimental feature for pruning the blockchain data as well, so enabling blob garbage collections is useful also there

**Review by fab-10 — APPROVED** [review]
LGTM.
It maybe be required to tune the following two options if the trielog purge is enabled and the space is not reclaimed quicky enough (cc @siladu )

- blob_garbage_collection_age_cutoff: the cutoff that the GC logic uses to determine which blob files should be considered “old.” For example, the default value of 0.25 signals to RocksDB that blobs residing in the oldest 25% of blob files should be relocated by GC. This parameter can be tuned to adjust the trade-off between write amplification and space amplification.

- blob_garbage_collection_force_threshold: if the ratio of garbage in the oldest blob files exceeds this threshold, targeted compactions are scheduled in order to force garbage collecting the blob files in question, assuming they are all eligible based on the value of blob_garbage_collection_age_cutoff above. This can help reduce space amplification in the case of skewed workloads where the affected files would not otherwise be picked up for compaction. This option is currently only supported with leveled compactions.

**Pull request merged**